// data-nfs/dist/src/v2/index.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
import xml2js from 'xml2js';
var version = require('../version');
var getNumber = function(item) {
    if (typeof item === 'undefined') return 0;
    if (typeof item === 'string') {
        var dotLength = item.split('').filter(function(i) {
            return i === '.';
        }).length;
        if (dotLength > 0 && item.indexOf(',') > -1) {
            var newItem = item.replace('.', '').replace(',', '.');
            return Number(newItem);
        }
        var newItem1 = item.replace(',', '.');
        return Number(newItem1);
    }
    return NaN;
};
var BASE_PRIMITIVES = {
    isCalculated: {
        value: false
    },
    isNumber: {
        value: false
    },
    isInt: {
        value: false
    }
};
var BASE_TEXT_ITEM = {
    value: '',
    isLike: true
};
var validatorFunctions = {
    isCalculated: function(foundItem, searchedItem, primitive) {
        var foundItemIsAObject = typeof foundItem === 'object';
        var searchedItemIsAObject = typeof searchedItem === 'object';
        return !foundItemIsAObject && !searchedItemIsAObject && primitive.isCalculated.value && getNumber(searchedItem) > getNumber(foundItem);
    },
    isNumber: function(searchedItem) {
        var searchedItemIsAObject = typeof searchedItem === 'object';
        return !searchedItemIsAObject && !isNaN(getNumber(searchedItem)) && !isNaN(parseFloat(getNumber(searchedItem).toString()));
    },
    isInt: function(searchedItem) {
        var searchedItemIsAObject = typeof searchedItem === 'object';
        return !searchedItemIsAObject && Number.isInteger(Number(searchedItem)) && (searchedItem === null || searchedItem === void 0 ? void 0 : searchedItem.length) === String(Number(searchedItem)).length;
    }
};
var validators = {
    cpnj: {
        text: [
            {
                value: 'cnpj',
                isLike: true
            },
            {
                value: 'documento',
                isLike: true
            },
            {
                value: 'CPF',
                isLike: false
            }
        ],
        keyValidatorsFrom: [
            {
                value: 'PrestadorServicos',
                isLike: false,
                mandatory: true
            },
            {
                value: 'emit',
                isLike: true,
                mandatory: true
            },
            {
                value: 'prest',
                isLike: true,
                mandatory: true
            },
            {
                value: 'pre',
                isLike: true,
                mandatory: true
            }
        ],
        keyValidatorsTo: [
            {
                value: 'TomadorServicos',
                isLike: false,
                mandatory: true
            },
            {
                value: 'toma',
                isLike: true,
                mandatory: true
            },
            {
                value: 'dest',
                isLike: true,
                mandatory: true
            },
            {
                value: 'tom',
                isLike: true,
                mandatory: true
            }
        ],
        primitive: BASE_PRIMITIVES
    },
    cpf: {
        text: [
            {
                value: 'cpf',
                isLike: true
            },
            {
                value: 'documento',
                isLike: true
            }
        ],
        keyValidatorsFrom: [
            {
                value: 'emit',
                isLike: true,
                mandatory: true
            },
            {
                value: 'prest',
                isLike: true,
                mandatory: true
            },
            {
                value: 'pre',
                isLike: true,
                mandatory: true
            }
        ],
        keyValidatorsTo: [
            {
                value: 'toma',
                isLike: true,
                mandatory: true
            },
            {
                value: 'dest',
                isLike: true,
                mandatory: true
            },
            {
                value: 'tom',
                isLike: true,
                mandatory: true
            }
        ],
        primitive: BASE_PRIMITIVES
    },
    value: {
        text: [
            {
                value: 'basecalc',
                isLike: true
            },
            {
                value: 'valorbase',
                isLike: true
            },
            {
                value: 'vServ',
                isLike: false
            },
            {
                value: 'valorservicos',
                isLike: false
            },
            {
                value: 'VALOR_SERVICO',
                isLike: false
            },
            {
                value: 'VlBasCalc',
                isLike: false
            },
            {
                value: 'VALOR_TOTAL',
                isLike: false
            },
            {
                value: 'VALORTOTALNOTAFISCAL',
                isLike: false
            }
        ],
        ignoredKeys: [
            {
                value: 'itens',
                isLike: true
            },
            {
                value: 'ITENSNOTA',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isCalculated: {
                value: true
            },
            isNumber: {
                value: true
            }
        })
    },
    grossValue: {
        text: [
            {
                value: 'vServ',
                isLike: false
            },
            {
                value: 'valorservicos',
                isLike: false
            },
            {
                value: 'VALOR_NOTA',
                isLike: false
            },
            {
                value: 'VALOR_TOTAL',
                isLike: false
            },
            {
                value: 'ValorNFS',
                isLike: false
            },
            {
                value: 'VALORTOTALNOTAFISCAL',
                isLike: false
            },
            {
                value: 'VALORTOTALBASEISSQN',
                isLike: false
            },
            {
                value: 'ValorTotal',
                isLike: false
            },
            {
                value: 'ValorTotalNota',
                isLike: false
            },
            {
                value: 'valorServico',
                isLike: false
            },
            {
                value: 'basecalc',
                isLike: true
            },
            {
                value: 'vprod',
                isLike: true
            },
            {
                value: 'valorbase',
                isLike: true
            }
        ],
        ignoredKeys: [
            {
                value: 'itens',
                isLike: true
            },
            {
                value: 'ITENSNOTA',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isCalculated: {
                value: true
            },
            isNumber: {
                value: true
            }
        })
    },
    liquidValue: {
        text: [
            {
                value: 'liquido',
                isLike: true
            },
            {
                value: 'Liquido',
                isLike: true
            },
            {
                value: 'vLiq',
                isLike: true
            },
            {
                value: 'VALORTOTALLIQUIDO',
                isLike: false
            }
        ],
        ignoredKeys: [
            {
                value: 'itens',
                isLike: true
            },
            {
                value: 'ITENSNOTA',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isCalculated: {
                value: true
            },
            isNumber: {
                value: true
            }
        })
    },
    number: {
        text: [
            {
                value: 'NUM_NOTA',
                isLike: false
            },
            {
                value: 'numero',
                isLike: false
            },
            {
                value: 'NumNF',
                isLike: false
            },
            {
                value: 'numero_nfse',
                isLike: false
            },
            {
                value: 'NOTAFISCALNUMERO',
                isLike: false
            },
            {
                value: 'NumeroNFe',
                isLike: false
            },
            {
                value: 'NumeroNFS-e',
                isLike: false
            },
            {
                value: 'NumeroNF',
                isLike: false
            },
            {
                value: 'cod',
                isLike: true
            },
            {
                value: 'nNF',
                isLike: true
            }
        ],
        ignoredKeys: [
            {
                value: 'emit',
                isLike: true,
                mandatory: true
            },
            {
                value: 'prest',
                isLike: true,
                mandatory: true
            },
            {
                value: 'toma',
                isLike: true,
                mandatory: true
            },
            {
                value: 'dest',
                isLike: true,
                mandatory: true
            },
            {
                value: 'tom',
                isLike: true,
                mandatory: true
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isNumber: {
                value: true
            }
        })
    },
    withheldIss: {
        text: [
            {
                value: 'issretido',
                isLike: true
            },
            {
                value: 'ISSQNRetido',
                isLike: true
            },
            {
                value: 'retido',
                isLike: false
            },
            {
                value: 'RETIDO',
                isLike: true
            },
            {
                value: 'Codigo',
                isLike: false
            },
            {
                value: 'ValorIssRet',
                isLike: false
            }
        ],
        ignoredKeys: [
            {
                value: 'itens',
                isLike: true
            },
            {
                value: 'servicos',
                isLike: true
            },
            {
                value: 'emit',
                isLike: true,
                mandatory: true
            },
            {
                value: 'prestador',
                isLike: true,
                mandatory: true
            },
            {
                value: 'toma',
                isLike: true,
                mandatory: true
            },
            {
                value: 'dest',
                isLike: true,
                mandatory: true
            },
            {
                value: 'tom',
                isLike: true,
                mandatory: true
            },
            {
                value: 'ITENSNOTA',
                isLike: false
            }
        ],
        keyValidators: [
            {
                value: 'ExigibilidadeISSQN',
                isLike: true,
                mandatory: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isNumber: {
                value: false
            }
        })
    },
    issValue: {
        text: [
            {
                value: 'valoriss',
                isLike: true
            },
            {
                value: 'VALOR_ISS',
                isLike: true
            },
            {
                value: 'ValorISSQNCalculado',
                isLike: true
            },
            {
                value: 'vISSQN',
                isLike: false
            },
            {
                value: 'VALOR_ISS_RET',
                isLike: false
            },
            {
                value: 'valorImposto',
                isLike: false
            },
            {
                value: 'IMPOSTO',
                isLike: false
            },
            {
                value: 'VALORTOTALISSQRETIDO',
                isLike: false
            },
            {
                value: 'ValorIssRetido',
                isLike: false
            },
            {
                value: 'Issqn',
                isLike: false
            }
        ],
        ignoredKeys: [
            {
                value: 'ITENSNOTA',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isNumber: {
                value: true
            },
            isCalculated: {
                value: true
            }
        })
    },
    inssValue: {
        text: [
            {
                value: 'inss',
                isLike: true
            },
            {
                value: 'INSS',
                isLike: true
            },
            {
                value: 'INSSRetido',
                isLike: true
            },
            {
                value: 'vINSS',
                isLike: false
            },
            {
                value: 'ValorINSS',
                isLike: false
            },
            {
                value: 'ValorINSSRetido',
                isLike: false
            },
            {
                value: 'VALOR_INSS',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isNumber: {
                value: true
            },
            isCalculated: {
                value: false
            }
        })
    },
    deductionValue: {
        text: [
            {
                value: 'deducoes',
                isLike: true
            },
            {
                value: 'vDesc',
                isLike: true
            },
            {
                value: 'Deduc',
                isLike: true
            },
            {
                value: 'vCalcDR',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isNumber: {
                value: true
            },
            isCalculated: {
                value: true
            }
        })
    },
    retentions: {
        text: [
            {
                value: 'retenc',
                isLike: true
            },
            {
                value: 'vTotalRet',
                isLike: true
            },
            {
                value: 'valorImposto',
                isLike: true
            },
            {
                value: 'retencao',
                isLike: true
            },
            {
                value: 'VALOR_IR',
                isLike: false
            },
            {
                value: 'ValorIr',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isNumber: {
                value: true
            },
            isCalculated: {
                value: true
            }
        })
    },
    operationNature: {
        text: [
            {
                value: 'NaturezaOperacao',
                isLike: true
            },
            {
                value: 'ExigibilidadeISS',
                isLike: true
            },
            {
                value: 'natureza',
                isLike: true
            },
            {
                value: 'operacao',
                isLike: true
            },
            {
                value: 'Codigo',
                isLike: false
            },
            {
                value: 'natOp',
                isLike: false
            }
        ],
        ignoredKeys: [
            {
                value: 'itens',
                isLike: true
            },
            {
                value: 'servicos',
                isLike: true
            }
        ],
        keyValidators: [
            {
                value: 'ExigibilidadeISSQN',
                isLike: true,
                mandatory: false
            }
        ],
        primitive: _object_spread({}, BASE_PRIMITIVES)
    },
    cnpjPrest: {
        text: [
            {
                value: 'CpfCnpjPre',
                isLike: false,
                notValidated: true
            },
            {
                value: 'CNPJCPFPrestador',
                isLike: false,
                notValidated: true
            }
        ],
        primitive: BASE_PRIMITIVES
    },
    cnpjToma: {
        text: [
            {
                value: 'CpfCnpjTom',
                isLike: false,
                notValidated: true
            },
            {
                value: 'CNPJCPFTomador',
                isLike: false,
                notValidated: true
            }
        ],
        primitive: BASE_PRIMITIVES
    },
    discount: {
        text: [
            {
                value: 'Desconto',
                isLike: true
            }
        ],
        ignoredKeys: [
            {
                value: 'itens',
                isLike: true
            },
            {
                value: 'ITENSNOTA',
                isLike: false
            }
        ],
        primitive: _object_spread_props(_object_spread({}, BASE_PRIMITIVES), {
            isCalculated: {
                value: true
            },
            isNumber: {
                value: true
            }
        })
    },
    emissionDate: {
        text: [
            {
                value: 'DataEmissao',
                isLike: true
            },
            {
                value: 'dhEmi',
                isLike: true
            },
            {
                value: 'data_nfse',
                isLike: false
            },
            {
                value: 'DATA',
                isLike: false
            }
        ],
        primitive: BASE_PRIMITIVES
    },
    emissionDateUnix: {
        text: [
            {
                value: 'iLocalMillis',
                isLike: false
            }
        ],
        primitive: BASE_PRIMITIVES,
        keyValidators: [
            {
                value: 'dataEmissao',
                isLike: false,
                mandatory: true
            }
        ]
    }
};
var findNf = function(ND) {
    var isArray = Array.isArray(ND);
    var isObject = typeof ND === 'object';
    if (isArray) {
        return findNf(ND[0]);
    }
    if (isObject) {
        var keys = Object.keys(ND);
        if (keys.includes('EnviarLoteRpsEnvio')) {
            throw new Error('Não é possivel validar o fiscal com o XML do RPS, é nescessario o XML da nota fiscal.');
        }
        if (keys.length > 1) {
            var indexValues = keys.findIndex(function(item) {
                return item.toLowerCase().indexOf('valor') > -1;
            });
            var indexNFSE = keys.findIndex(function(item) {
                return item.toLowerCase().indexOf('nfse') > -1;
            });
            var indexNFE = keys.findIndex(function(item) {
                return item.toLowerCase().indexOf('nfe') > -1;
            });
            if (indexNFSE > -1 && keys.length < 7 && indexValues < 0) {
                return findNf(ND[keys[indexNFSE]]);
            }
            if (indexNFE > -1 && keys.length < 7 && indexValues < 0) {
                return findNf(ND[keys[indexNFE]]);
            }
            return ND;
        }
        return findNf(ND[keys[0]]);
    }
    return null;
};
var searchItem = function(NF, item, validator, foundItem) {
    var primitive = arguments.length > 4 && arguments[4] !== void 0 ? arguments[4] : BASE_PRIMITIVES, ignoredKeys = arguments.length > 5 && arguments[5] !== void 0 ? arguments[5] : [];
    var _findItem;
    var searchedItem = (_findItem = findItem(NF, item, validator, ignoredKeys)) === null || _findItem === void 0 ? void 0 : _findItem.found;
    var foundItemIsAObject = typeof foundItem === 'object';
    var searchedItemIsAObject = typeof searchedItem === 'object';
    var hasSearchedItemAndFoundItemIsNotValid = (!foundItem || foundItemIsAObject) && searchedItem;
    var isCalculatedAndIsHigher = validatorFunctions.isCalculated(foundItem, searchedItem, primitive);
    var isValid = hasSearchedItemAndFoundItemIsNotValid || isCalculatedAndIsHigher;
    var isNumber = validatorFunctions.isNumber(searchedItem);
    if (primitive.isNumber.value) {
        var isInt = validatorFunctions.isInt(searchedItem);
        if (isValid && isNumber) {
            return primitive.isInt.value ? isInt ? searchedItem : undefined : searchedItem;
        }
        return;
    }
    if (isValid) {
        var _findItem1;
        var filteredItem = searchedItemIsAObject ? (_findItem1 = findItem(searchedItem, item, undefined, ignoredKeys)) === null || _findItem1 === void 0 ? void 0 : _findItem1.found : searchedItem;
        return filteredItem;
    }
    return undefined;
};
var findItemByList = function(NF, list) {
    var validatorsList = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : [], primitive = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : BASE_PRIMITIVES, ignoredKeys = arguments.length > 4 && arguments[4] !== void 0 ? arguments[4] : [];
    var foundItem;
    var keys = Object.keys(NF);
    list.forEach(function(item) {
        if (validatorsList.length > 0 && !item.notValidated) {
            var shouldReturn = false;
            validatorsList.forEach(function(valid) {
                if (keys.length < 5) {
                    keys.forEach(function(key) {
                        var nf = Array.isArray(NF[key]) ? NF[key][0] : NF[key];
                        var validKey = key.toLowerCase().indexOf(valid.value.toLowerCase()) > -1;
                        var searchedItem = searchItem(nf, item, validKey ? undefined : valid, foundItem, primitive, ignoredKeys);
                        if (searchedItem) {
                            foundItem = searchedItem;
                        }
                    });
                } else {
                    var searchedItem = searchItem(NF, item, valid, foundItem, primitive, ignoredKeys);
                    if (searchedItem) {
                        foundItem = searchedItem;
                    }
                }
                if (!shouldReturn) {
                    shouldReturn = !!valid.mandatory;
                }
            });
            if (shouldReturn) {
                return;
            }
        }
        var searchedItem = searchItem(NF, item, BASE_TEXT_ITEM, foundItem, primitive, ignoredKeys);
        if (searchedItem) {
            foundItem = searchedItem;
        }
    });
    if (typeof foundItem === 'object') {
        var values = Object.values(foundItem);
        return Array.isArray(values[0]) ? values[0][0] : values[0];
    }
    return foundItem;
};
var findItem = function(NF, item) {
    var keyValidator = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : BASE_TEXT_ITEM, ignoredKeys = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : [];
    if (keyValidator.value.trim()) {
        var keys = Object.keys(NF);
        var filteredKeys = keys.filter(function(key) {
            return key.toLowerCase().indexOf(keyValidator.value.toLowerCase()) > -1;
        });
        if (filteredKeys.length > 5) {
            var keyIndex = filteredKeys.findIndex(function(filtKey) {
                return filtKey.toLowerCase().indexOf(item.value.toLowerCase()) > -1;
            });
            var value = findVal(NF, _object_spread_props(_object_spread({}, item), {
                value: filteredKeys[keyIndex]
            }), undefined, undefined, ignoredKeys);
            if (value) {
                return {
                    found: value
                };
            }
        }
    }
    var value1 = findVal(NF, item, keyValidator.value.trim() ? keyValidator : undefined, undefined, ignoredKeys);
    if (value1) {
        return {
            found: value1
        };
    }
    return null;
};
function findVal(object, keyObj, keyValidator) {
    var validated = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : false, ignoredKeys = arguments.length > 4 && arguments[4] !== void 0 ? arguments[4] : [];
    var value;
    Object.keys(object).some(function(k) {
        if (keyObj.value === undefined) return false;
        if (ignoredKeys.length > 0) {
            var index = ignoredKeys.findIndex(function(item) {
                return item.isLike ? k.toLowerCase().indexOf(item.value.toLowerCase()) > -1 : k.toLowerCase() === item.value.toLowerCase();
            });
            if (index > -1) return false;
        }
        if (keyValidator && (keyValidator.isLike ? k.toLowerCase().indexOf(keyValidator.value.toLowerCase()) > -1 : k.toLowerCase() === keyValidator.value.toLowerCase())) {
            var kValue = Array.isArray(object[k]) ? object[k][0] : object[k];
            value = findVal(object[k], keyObj, keyValidator, typeof kValue === 'object', ignoredKeys);
            return value !== undefined;
        }
        if (keyObj.isLike ? k.toLowerCase().indexOf(keyObj.value.toLowerCase()) > -1 : k.toLowerCase() === keyObj.value.toLowerCase()) {
            if (keyValidator && validated || !keyValidator) {
                value = Array.isArray(object[k]) ? object[k][0] : object[k];
                return true;
            }
        }
        if (object[k] && typeof object[k] === 'object') {
            value = findVal(object[k], keyObj, keyValidator, validated, ignoredKeys);
            return value !== undefined;
        }
    });
    return value;
}
var getDataNFSv2 = function() {
    var _ref = _async_to_generator(function(xmlString) {
        var stripPrefix, result, nfs, cnpjEmitWithValidator, cnpjPrest, cnpjToma, cnpjDestWithValidator, cpfEmit, cnpjEmit, cnpjDest, realCnpjEmit, realCnpjDest, realCpfEmit, valorNF, numero, _findItemByList, iss, _findItemByList1, valorIss, valorBruto, valorLiquido, _findItemByList2, valorDeducoes, _findItemByList3, retencoes, descontos, naturezaOperacao, inssValue, emissionDate, emissionDateUnix, isBrazilianDate, dateText, dates, dataEmissao, valorLiquidoValidado, valorBrutoValidado;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    console.log('running data-nf@' + version.LIB_VERSION);
                    stripPrefix = xml2js.processors.stripPrefix;
                    return [
                        4,
                        new Promise(function(resolve, reject) {
                            var parser = new xml2js.Parser({
                                attrkey: 'ATTR',
                                tagNameProcessors: [
                                    stripPrefix
                                ],
                                attrNameProcessors: [
                                    stripPrefix
                                ]
                            });
                            parser.parseString(xmlString, function(error, result) {
                                if (error) resolve(false);
                                else resolve(result);
                            });
                        })
                    ];
                case 1:
                    result = _state.sent();
                    if (result === null || result === void 0 ? void 0 : result.EnviarLoteRpsEnvio) {
                        throw Error('Não é possivel validar o fiscal com o XML do RPS, é nescessario o XML da nota fiscal.');
                    }
                    try {
                        nfs = findNf(result);
                        cnpjEmitWithValidator = findItemByList(nfs, validators.cpnj.text, validators.cpnj.keyValidatorsFrom);
                        cnpjPrest = findItemByList(nfs, validators.cnpjPrest.text);
                        cnpjToma = findItemByList(nfs, validators.cnpjToma.text);
                        cnpjDestWithValidator = findItemByList(nfs, validators.cpnj.text, validators.cpnj.keyValidatorsTo);
                        cpfEmit = findItemByList(nfs, validators.cpf.text, validators.cpf.keyValidatorsFrom);
                        cnpjEmit = cnpjEmitWithValidator ? cnpjEmitWithValidator : cnpjPrest ? cnpjPrest : undefined;
                        cnpjDest = cnpjDestWithValidator ? cnpjDestWithValidator : cnpjToma ? cnpjToma : undefined;
                        realCnpjEmit = cnpjEmit ? cnpjEmit.replace(/\D/g, '') : undefined;
                        realCnpjDest = cnpjDest ? cnpjDest.replace(/\D/g, '') : undefined;
                        realCpfEmit = cpfEmit ? cpfEmit.replace(/\D/g, '') : undefined;
                        valorNF = findItemByList(nfs, validators.value.text, [], validators.value.primitive);
                        numero = findItemByList(nfs, validators.number.text, [], validators.number.primitive, validators.number.ignoredKeys);
                        ;
                        iss = (_findItemByList = findItemByList(nfs, validators.withheldIss.text, [], validators.withheldIss.primitive, validators.withheldIss.ignoredKeys)) !== null && _findItemByList !== void 0 ? _findItemByList : '0';
                        ;
                        valorIss = (_findItemByList1 = findItemByList(nfs, validators.issValue.text, [], validators.issValue.primitive, validators.issValue.ignoredKeys)) !== null && _findItemByList1 !== void 0 ? _findItemByList1 : '0';
                        valorBruto = findItemByList(nfs, validators.grossValue.text, [], validators.grossValue.primitive, validators.grossValue.ignoredKeys);
                        valorLiquido = findItemByList(nfs, validators.liquidValue.text, [], validators.liquidValue.primitive, validators.liquidValue.ignoredKeys);
                        ;
                        valorDeducoes = (_findItemByList2 = findItemByList(nfs, validators.deductionValue.text, [], validators.deductionValue.primitive)) !== null && _findItemByList2 !== void 0 ? _findItemByList2 : '0';
                        ;
                        retencoes = (_findItemByList3 = findItemByList(nfs, validators.retentions.text, [], validators.retentions.primitive)) !== null && _findItemByList3 !== void 0 ? _findItemByList3 : '0';
                        descontos = findItemByList(nfs, validators.discount.text, [], validators.discount.primitive, validators.discount.ignoredKeys);
                        naturezaOperacao = findItemByList(nfs, validators.operationNature.text, validators.operationNature.keyValidators, validators.operationNature.primitive, validators.operationNature.ignoredKeys);
                        inssValue = findItemByList(nfs, validators.inssValue.text, [], validators.inssValue.primitive, validators.inssValue.ignoredKeys);
                        emissionDate = findItemByList(nfs, validators.emissionDate.text, [], validators.emissionDate.primitive);
                        emissionDateUnix = findItemByList(nfs, validators.emissionDateUnix.text, validators.emissionDateUnix.keyValidators, validators.emissionDateUnix.primitive);
                        isBrazilianDate = emissionDate === null || emissionDate === void 0 ? void 0 : emissionDate.match(/[0-9]{2}\/[0-9]{2}\/[0-9]{4}/);
                        dateText = '';
                        if (isBrazilianDate && emissionDate) {
                            dates = emissionDate.split('/');
                            dateText = "".concat(dates[2], "-").concat(dates[1], "-").concat(dates[0], " 04:00:00");
                        } else if (emissionDate) {
                            dateText = emissionDate;
                        }
                        dataEmissao = emissionDateUnix ? new Date(Number(emissionDateUnix)) : emissionDate ? new Date(dateText) : undefined;
                        valorLiquidoValidado = !!valorLiquido ? getNumber(valorLiquido) : getNumber(valorNF) - getNumber(retencoes);
                        valorBrutoValidado = descontos ? getNumber(valorBruto) - getNumber(descontos) >= getNumber(valorLiquido) ? getNumber(valorBruto) - getNumber(descontos) : getNumber(valorBruto) : getNumber(valorBruto);
                        return [
                            2,
                            {
                                cnpjEmit: realCnpjEmit,
                                cnpjDest: realCnpjDest,
                                cpfEmit: realCpfEmit,
                                retencoes: retencoes,
                                valorDeducoes: valorDeducoes,
                                valorIss: valorIss,
                                valorBruto: String(valorBrutoValidado),
                                valorLiquido: String(valorLiquidoValidado),
                                valorNF: valorNF,
                                inss: inssValue,
                                descontos: descontos,
                                iss: iss,
                                numero: numero,
                                naturezaOperacao: naturezaOperacao,
                                dataEmissao: dataEmissao
                            }
                        ];
                    } catch (error) {
                        throw error;
                    }
                    return [
                        2
                    ];
            }
        });
    });
    return function getDataNFSv2(xmlString) {
        return _ref.apply(this, arguments);
    };
}();
export default getDataNFSv2;
