// data-nfs/dist/src/legado/index.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
var xml2js = require('xml2js');
var getDataNFSLegado = function() {
    var _ref = _async_to_generator(function(xmlString) {
        var stripPrefix, result, nf, layout, _result_GerarNfseResposta, _result_EnviarLoteRpsSincronoResposta, _result_LoteNotaFiscal, _result_ConsultarNfseLote, _result_ConsultarNfseResposta, _result_ConsultarNfseFaixaResposta, _result_ConsultarNfseRpsResposta_CompNfse_, _result_ConsultarNfseRpsResposta, _result_NFSE, _result_ConsultarNfseServicoPrestadoResposta, _result_GerarNfseResposta1, _result_EnviarLoteRpsSincronoResposta1, _result_LoteNotaFiscal1, _result_ConsultarNfseServicoPrestadoResposta1, dados, cnpjEmit, cnpjDest, valorNF, Numero;
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    stripPrefix = xml2js.processors.stripPrefix;
                    return [
                        4,
                        new Promise(function(resolve, reject) {
                            var parser = new xml2js.Parser({
                                attrkey: "ATTR",
                                tagNameProcessors: [
                                    stripPrefix
                                ],
                                attrNameProcessors: [
                                    stripPrefix
                                ]
                            });
                            parser.parseString(xmlString, function(error, result) {
                                if (error) resolve(false);
                                else resolve(result);
                            });
                        })
                    ];
                case 1:
                    result = _state.sent();
                    //  LAYOUT ABRASF contem ConsultarNfseResposta
                    //  LAYOUT DIVERGENTE 1 contem ConsultarNfseRpsResposta
                    if (result === null || result === void 0 ? void 0 : result.EnviarLoteRpsEnvio) {
                        return [
                            2,
                            [
                                false,
                                'Não é possivel validar o fiscal com o XML do RPS, é nescessario o XML da nota fiscal.'
                            ]
                        ];
                    }
                    try {
                        ;
                        if (result === null || result === void 0 ? void 0 : (_result_GerarNfseResposta = result.GerarNfseResposta) === null || _result_GerarNfseResposta === void 0 ? void 0 : _result_GerarNfseResposta.ListaNfse) {
                            ;
                            nf = (result === null || result === void 0 ? void 0 : (_result_GerarNfseResposta1 = result.GerarNfseResposta) === null || _result_GerarNfseResposta1 === void 0 ? void 0 : _result_GerarNfseResposta1.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0]) || false;
                        } else if (result === null || result === void 0 ? void 0 : result.ComplNfse) {
                            nf = (result === null || result === void 0 ? void 0 : result.ComplNfse.Nfse[0].InfNfse[0]) || false;
                        } else if (result === null || result === void 0 ? void 0 : (_result_EnviarLoteRpsSincronoResposta = result.EnviarLoteRpsSincronoResposta) === null || _result_EnviarLoteRpsSincronoResposta === void 0 ? void 0 : _result_EnviarLoteRpsSincronoResposta.ListaNfse) {
                            ;
                            // console.log(result.ConsultarNfseLote.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0])
                            nf = (result === null || result === void 0 ? void 0 : (_result_EnviarLoteRpsSincronoResposta1 = result.EnviarLoteRpsSincronoResposta) === null || _result_EnviarLoteRpsSincronoResposta1 === void 0 ? void 0 : _result_EnviarLoteRpsSincronoResposta1.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0]) || false;
                        } else if (result === null || result === void 0 ? void 0 : result.Nfse) {
                            // console.log(result?.Nfse.Numero)
                            nf = (result === null || result === void 0 ? void 0 : result.Nfse) || false;
                        } else if (result === null || result === void 0 ? void 0 : (_result_LoteNotaFiscal = result.LoteNotaFiscal) === null || _result_LoteNotaFiscal === void 0 ? void 0 : _result_LoteNotaFiscal.CompNfse) {
                            ;
                            // console.log(result.ConsultarNfseLote.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0])
                            nf = (result === null || result === void 0 ? void 0 : (_result_LoteNotaFiscal1 = result.LoteNotaFiscal) === null || _result_LoteNotaFiscal1 === void 0 ? void 0 : _result_LoteNotaFiscal1.CompNfse[0].Nfse[0].InfNfse[0]) || false;
                        } else if (result === null || result === void 0 ? void 0 : (_result_ConsultarNfseLote = result.ConsultarNfseLote) === null || _result_ConsultarNfseLote === void 0 ? void 0 : _result_ConsultarNfseLote.ListaNfse) {
                            // console.log(result.ConsultarNfseLote.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0])
                            nf = result.ConsultarNfseLote.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0] || false;
                        } else if (typeof result.CompNfse !== 'undefined') {
                            if (typeof result.CompNfse.Nfse !== 'undefined') {
                                nf = result.CompNfse.Nfse[0].InfNfse[0] || false;
                            } else if (result.CompNfse[0] !== 'undefined') {
                                nf = result.CompNfse[0].Nfse[0].InfNfse[0] || false;
                            } else {
                                nf = false;
                            }
                        } else if (typeof (result === null || result === void 0 ? void 0 : (_result_ConsultarNfseResposta = result.ConsultarNfseResposta) === null || _result_ConsultarNfseResposta === void 0 ? void 0 : _result_ConsultarNfseResposta.ListaNfse) !== 'undefined') {
                            nf = result.ConsultarNfseResposta.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0] || false;
                        } else if (typeof (result === null || result === void 0 ? void 0 : (_result_ConsultarNfseFaixaResposta = result.ConsultarNfseFaixaResposta) === null || _result_ConsultarNfseFaixaResposta === void 0 ? void 0 : _result_ConsultarNfseFaixaResposta.ListaNfse) !== 'undefined') {
                            nf = result.ConsultarNfseFaixaResposta.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0] || false;
                        } else if (typeof (result === null || result === void 0 ? void 0 : (_result_ConsultarNfseRpsResposta = result.ConsultarNfseRpsResposta) === null || _result_ConsultarNfseRpsResposta === void 0 ? void 0 : (_result_ConsultarNfseRpsResposta_CompNfse_ = _result_ConsultarNfseRpsResposta.CompNfse[0]) === null || _result_ConsultarNfseRpsResposta_CompNfse_ === void 0 ? void 0 : _result_ConsultarNfseRpsResposta_CompNfse_.Nfse) !== 'undefined') {
                            nf = (result === null || result === void 0 ? void 0 : result.ConsultarNfseRpsResposta.CompNfse[0].Nfse[0].InfNfse[0]) || false;
                        } else if (typeof (result === null || result === void 0 ? void 0 : (_result_NFSE = result.NFSE) === null || _result_NFSE === void 0 ? void 0 : _result_NFSE.NOTA) !== 'undefined') {
                            nf = result.NFSE.NOTA[0] || false;
                        } else if (typeof (result === null || result === void 0 ? void 0 : result.nota) !== 'undefined') {
                            // console.log('###RESULT#nota', result)
                            nf = result.nota || false;
                        } else if (_type_of(result === null || result === void 0 ? void 0 : (_result_ConsultarNfseServicoPrestadoResposta = result.ConsultarNfseServicoPrestadoResposta) === null || _result_ConsultarNfseServicoPrestadoResposta === void 0 ? void 0 : _result_ConsultarNfseServicoPrestadoResposta.ListaNfse)) {
                            ;
                            // console.log('ConsultarNfseServicoPrestadoResposta', result?.ConsultarNfseServicoPrestadoResposta)
                            nf = (result === null || result === void 0 ? void 0 : (_result_ConsultarNfseServicoPrestadoResposta1 = result.ConsultarNfseServicoPrestadoResposta) === null || _result_ConsultarNfseServicoPrestadoResposta1 === void 0 ? void 0 : _result_ConsultarNfseServicoPrestadoResposta1.ListaNfse[0].CompNfse[0].Nfse[0].InfNfse[0]) || false;
                        } else {
                            return [
                                2,
                                [
                                    false,
                                    'LAYOUT DE XML DESCONHECIDO, ENTRE EM CONTATO COM O SUPORTE.'
                                ]
                            ];
                        }
                    } catch (error) {
                        nf = false;
                    }
                    if (!nf) {
                        return [
                            2,
                            [
                                false,
                                "LAYOUT DE XML DESCONHECIDO, ENTRE EM CONTATO COM O SUPORTE."
                            ]
                        ];
                    }
                    dados = false;
                    cnpjEmit = false;
                    cnpjDest = false;
                    valorNF = false;
                    Numero = false;
                    //  ######### ------------  cnpjEmit
                    try {
                        cnpjEmit = nf.PrestadorServico[0].IdentificacaoPrestador[0].CpfCnpj[0].Cnpj[0];
                    } catch (error) {}
                    if (!cnpjEmit) {
                        try {
                            cnpjEmit = nf.PrestadorServico[0].IdentificacaoPrestador[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjEmit) {
                        try {
                            cnpjEmit = nf.DeclaracaoPrestacaoServico[0].Prestador[0].IdentificacaoPrestador[0].CpfCnpj[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjEmit) {
                        try {
                            cnpjEmit = nf.DeclaracaoPrestacaoServico[0].IdentificacaoPrestador[0].CpfCnpj[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjEmit) {
                        try {
                            cnpjEmit = nf.DeclaracaoPrestacaoServico[0].InfDeclaracaoPrestacaoServico[0].Prestador[0].CpfCnpj[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjEmit) {
                        try {
                            cnpjEmit = nf.PRESTCPFCNPJ[0];
                        } catch (error) {}
                    }
                    if (!cnpjEmit) {
                        try {
                            cnpjEmit = nf.prestador[0].cpfcnpj[0];
                        } catch (error) {}
                    }
                    //  ######### ------------  cnpjDest
                    try {
                        cnpjDest = nf.TomadorServico[0].IdentificacaoTomador[0].CpfCnpj[0].Cnpj[0];
                    } catch (error) {}
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.TomadorServico[0].IdentificacaoTomador[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.DeclaracaoPrestacaoServico[0].Tomador[0].IdentificacaoTomador[0].CpfCnpj[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.DeclaracaoPrestacaoServico[0].InfDeclaracaoPrestacaoServico[0].Tomador[0].IdentificacaoTomador[0].CpfCnpj[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.DeclaracaoPrestacaoServico[0].InfDeclaracaoPrestacaoServico[0].Tomador[0].IdentificacaoTomador[0].CpfCnpj[0].Cpf[0];
                        } catch (error) {}
                    }
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.DeclaracaoPrestacaoServico[0].InfDeclaracaoPrestacaoServico[0].TomadorServico[0].IdentificacaoTomador[0].CpfCnpj[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.DeclaracaoPrestacaoServico[0].DadosTomador[0].IdentificacaoTomador[0].CpfCnpj[0].Cnpj[0];
                        } catch (error) {}
                    }
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.TOMCPFCNPJ[0];
                        } catch (error) {}
                    }
                    if (!cnpjDest) {
                        try {
                            cnpjDest = nf.tomador[0].cpfcnpj[0];
                        } catch (error) {}
                    }
                    //  ######### ------------  valorNF
                    try {
                        valorNF = nf.Servico[0].Valores[0].ValorServicos[0];
                    } catch (error) {}
                    if (!valorNF) {
                        try {
                            valorNF = nf.ValoresNfse[0].BaseCalculo[0];
                        } catch (error) {}
                    }
                    if (!valorNF) {
                        try {
                            valorNF = nf.DeclaracaoPrestacaoServico[0].ValorServicos[0];
                        } catch (error) {}
                    }
                    if (!valorNF) {
                        try {
                            valorNF = nf.DeclaracaoPrestacaoServico[0].InfDeclaracaoPrestacaoServico[0].Servico[0].Valores[0].ValorServicos[0];
                        } catch (error) {}
                    }
                    if (!valorNF) {
                        try {
                            valorNF = nf.DeclaracaoPrestacaoServico[0].InfDeclaracaoPrestacaoServico[0].Servico[0].Valores[0].ValorServicos[0];
                        } catch (error) {}
                    }
                    if (!valorNF) {
                        try {
                            valorNF = nf.VALORTOTALSERVICOS[0];
                        } catch (error) {}
                    }
                    if (!valorNF) {
                        try {
                            valorNF = nf.basecalculo[0];
                        } catch (error) {}
                    }
                    //  ######### ------------  chave
                    try {
                        Numero = nf.Numero[0];
                    } catch (error) {}
                    if (!Numero) {
                        try {
                            Numero = nf.COD[0];
                        } catch (error) {}
                    }
                    if (!Numero) {
                        try {
                            Numero = nf.numero[0];
                        } catch (error) {}
                    }
                    if (cnpjEmit !== false) {
                        cnpjEmit = String(cnpjEmit).replace(/[^0-9]/g, '');
                    }
                    if (cnpjDest !== false) {
                        cnpjDest = String(cnpjDest).replace(/[^0-9]/g, '');
                    }
                    // console.log({
                    //   cnpjEmit,
                    //   cnpjDest,
                    //   valorNF,
                    //   Numero
                    // })
                    if (cnpjEmit && cnpjDest && valorNF && Numero) {
                        dados = {
                            cnpjEmit: cnpjEmit,
                            cnpjDest: cnpjDest,
                            valorNF: valorNF,
                            Numero: Numero
                        };
                    } else {
                        return [
                            2,
                            [
                                false,
                                "ERRO AO LER DADOS DO DESTE LAYOUT, ENTRE EM CONTATO COM O SUPORTE.\n      ".concat(String(Boolean(cnpjEmit)), " | \n      ").concat(String(Boolean(cnpjDest)), " | \n      ").concat(String(Boolean(valorNF)), " | \n      ").concat(String(Boolean(Numero)), " | \n    ")
                            ]
                        ];
                    }
                    return [
                        2,
                        [
                            true,
                            dados
                        ]
                    ];
            }
        });
    });
    return function getDataNFSLegado(xmlString) {
        return _ref.apply(this, arguments);
    };
}();
export default getDataNFSLegado;
